File: clojure/log-levels/src/log_levels.clj
(ns log-levels
  (:require [clojure.string :as str]))

(def pattern 
  (re-pattern "\\[(?<level>INFO|WARNING|ERROR)\\]:\\s+(?<message>[a-zA-Z0-9 ]+)")
  )

(defn matcher
  [s]
  (re-matcher pattern s)
  )

(defn re-group
  [s group-name]
  (let [
        matcher (matcher s)
        finds (re-find matcher)
        ]
      (.group matcher group-name)
    )
  )

(defn message
  "Takes a string representing a log line
   and returns its message with whitespace trimmed."
  [s]
  (str/trim (re-group s "message"))
  )

(defn log-level
  "Takes a string representing a log line
   and returns its level in lower-case."
  [s]
  (str/lower-case (re-group s "level"))
  )

(defn reformat
  "Takes a string representing a log line and formats it
   with the message first and the log level in parentheses."
  [s]
  (str (message s) " (" (log-level s) ")")
  )

File: cpp/log-levels/log_levels.cpp
#include <string>
#include <regex>

using namespace std;

namespace log_line {
    regex pattern() {
        regex rgx("\\[(INFO|WARNING|ERROR)\\]: +([a-zA-Z0-9 ]+)");
        return rgx;
    }

    string message(string log) {
        return regex_replace(log, log_line::pattern(), "$2");
    }

    string log_level(string log) {
        return regex_replace(log, log_line::pattern(), "$1");
    }

    string reformat(string log) {
        return regex_replace(log, log_line::pattern(), "$2 ($1)");
    }
} // namespace log_line
